wikipedia_cleanup/data_filter.py

```diff
@@ -0,0 +1,266 @@
+            f"Filtered By current Filter: \t\t {num_self_deletions} "
+            f"\t current:\t {num_self_deletions / self.input_num_changes * 100} %"
+            f"\t total:\t {num_self_deletions / self.initial_num_changes * 100} %\n"
+        )
+
+    def add_stats(self, other_stats: "FilterStats") -> None:
+        self.initial_num_changes += other_stats.initial_num_changes
+        self.input_num_changes += other_stats.input_num_changes
+        self.output_num_changes += other_stats.output_num_changes
+
+
+class AbstractDataFilter(ABC):
+    _filter_stats: FilterStats
+
+    def __init__(self) -> None:
+        self._filter_stats = FilterStats()
+
+    def filter(
+        self, changes: List[InfoboxChange], initial_num_changes: int
+    ) -> List[InfoboxChange]:
+        if self._filter_stats.initial_num_changes != INITIAL_STATS_VALUE:
+            print(
+                "WARNING: Using a filter where the stats are not resetted. "
+                "Thus the stats will be overwritten."
+            )
+        self._filter_stats.initial_num_changes = initial_num_changes
+        self._filter_stats.input_num_changes = len(changes)
+        filtered_changes = []
+        start_idx = 0
+        for end_idx in range(len(changes)):
+            if (
+                changes[start_idx].infobox_key != changes[end_idx].infobox_key
+                or changes[start_idx].property_name != changes[end_idx].property_name
+            ):
+                filtered_changes.extend(
+                    self._filter_for_property(changes[start_idx:end_idx])
+                )
+                start_idx = end_idx
+        self._filter_stats.output_num_changes = len(filtered_changes)
+        return filtered_changes
+
+    @abstractmethod
+    def _filter_for_property(self, changes: List[InfoboxChange]) -> List[InfoboxChange]:
+        pass
+
+    @property
+    def filter_stats(self) -> FilterStats:
+        return self._filter_stats
+
+    def __str__(self) -> str:
+        base_print_width = 30
+        return (
+            f'{"+" * base_print_width}\n'
+            f"{self.__class__.__name__}\n"
+            f'{"+" * base_print_width}\n' + str(self.filter_stats)
+        )
+
+
+class DataFilterMinNumChanges(AbstractDataFilter):
+    _min_number_of_changes: int
+
+    def __init__(self, min_number_of_changes: int = 5):
+        super().__init__()
+        self._min_number_of_changes = min_number_of_changes
+
+    @property
+    def min_number_of_changes(self) -> int:
+        return self._min_number_of_changes
+
+    def _filter_for_property(self, changes: List[InfoboxChange]) -> List[InfoboxChange]:
+        return [] if len(changes) < self._min_number_of_changes else changes
+
+
+class DataFilterMajorityValuePerDay(AbstractDataFilter):
+    """
+    This filter needs to process each tuple (infobox, property_name, day).
+    Therefore the method is overwritten / slightly changes.
+    """
```

Comment:
The generic filter passes `(Infobox, property_name)` which includes all information for the days. Therefore, you could do the per-day check inside the `_filter_for_property` method and avoid overwriting the `filter` method

Reply:
Jup I thought about that as well but doing it this way saves one loop over the data. I think this is faster and thats why I did it this way